### DISCUSSION TITLE
Maybe is there a mistake in the vocabulary?

### DISCUSSION THREAD

**Discussion by lorenzosquadrani** [brianhie/viral-mutation#22] [Q&A] [answered]
Hello,

First let me congratulate you. This is a very interesting approach to viral escape.

There is something I don't understand, which may be a mistake.
In the file flu.py, line 237:
` AAs = [  'A', 'R', 'N', 'D', 'C', 'Q', 'E', 'G', 'H', 'I', 'L', 'K', 'M', 'F', 'P', 'S', 'T', 'W','Y', 'V', 'X', 'Z', 'J', 'U', 'B', 'Z' ]`
`vocabulary = { aa: idx + 1 for idx, aa in enumerate(sorted(AAs)) }`
The letter 'O' is missing from the list. Besides the letter 'Z' appears two times (not influencing the dictionary however).

Is there a reason for the 'O' missing?

Thank you,
Lorenzo

**Comment by brianhie** [comment, marked as answer]
Hi @lorenzosquadrani, the amino acid "O" (pyrrolysine) is a non-standard amino acid and so we did not consider it. Pyrrolysine is not used by viruses or humans and I believe is only used by certain kinds of bacteria.

**Reply by lorenzosquadrani** [reply]
Got it, thanks (sorry for the silly question, I know nothing about biology).

Still, I think something may be wrong.
Again in the file **flu.py**, in the _setup_ function, you set _vocab_size=len(AAs)_. But _len(AAs)=26_, while the number of letters in the vocabulary is 25 without the "O" (the reason why AAs contains 26 elements is because "Z" appears two times).
I know this objections may seem silly, but indeed they affect the number of units in the last Dense layer, affecting maybe the interpretation of its output.

Have a nice day!

**Reply by brianhie** [reply]
Hi @lorenzosquadrani, unfortunately just in `flu.py` there is a slight issue where both the "start" token and "Z" will be assigned 26 in the lookup table (end character is 27, pad is 0). Fortunately, "Z" does not correspond to a real amino acid, so it should only appear very rarely (or at all) in the flu training data. And it shouldn't affect the results, since we [ignore "Z"](https://github.com/brianhie/viral-mutation/blob/6c463b1744038a6d3e5927fe8a3270f9cd51fcbc/bin/cached_semantics.py#L66) when computing the final values, as "Z" is not a natural amino acid.
